plugins/atomicfu/atomicfu-compiler/src/org.jetbrains.kotlinx.atomicfu.compiler/extensions/AtomicfuTransformer.kt

```diff
@@ -101,6 +103,44 @@ class AtomicfuTransformer(private val context: IrPluginContext) {
 
     private inner class AtomicTransformer : IrElementTransformer<IrFunction?> {
 
+        override fun visitProperty(declaration: IrProperty, data: IrFunction?): IrStatement {
+            // Support transformation for delegated properties:
+            if (declaration.isDelegated && declaration.backingField?.type?.isAtomicValueType() == true) {
+                declaration.backingField?.let { delegateBackingField ->
+                    delegateBackingField.initializer?.let {
+                        val initializer = it.expression as IrCall
+                        when {
+                            initializer.isAtomicFieldGetter() -> {
+                                // val _a = atomic(0)
+                                // var a: Int by _a
+                                // Accessors of the delegated property `a` are implemented via the generated property `a$delegate`,
+                                // that is the copy of the original `_a`.
+                                // They should be delegated to the value of the original field `_a` instead of `a$delegate`.
+
+                                // fun <get-a>() = a$delegate.value -> _a.value
+                                // fun <set-a>(value: Int) = { a$delegate.value = value } -> { _a.value = value }
+                                val initializerName = initializer.symbol.owner.name.asString()
+                                val originalFieldName = initializerName.getFieldName()
+                                val originalProperty = (declaration.parent as IrDeclarationContainer).declarations.single {
+                                    it is IrProperty && it.name.asString() == originalFieldName &&
+                                            it.backingField != null && it.backingField!!.type == delegateBackingField.type
+                                }
+                                declaration.transform(DelegatePropertyTransformer((originalProperty as IrProperty).backingField!!), null)
+                            }
+                            initializer.isAtomicFactory() -> {
+                                // var a by atomic(77) -> var a: Int = 77
+                                it.expression = initializer.eraseAtomicFactory()
+                                    ?: error("Atomic factory was expected but found ${initializer.render()}")
+                                declaration.transform(DelegatePropertyTransformer(delegateBackingField), null)
+                            }
+                            else -> error("Unexpected initializer of the delegated property: $initializer")
+                        }
+                    }
+                }
+            }
```

Comment:
I think modifying accessors here would be a bit more straightforward and simple instead of doing so in the separate lowering through backings field